alf/algorithms/sac_algorithm.py

```diff
@@ -623,35 +655,45 @@ def _select_q_value(self, action, q_values):
         """Use ``action`` to index and select Q values.
         Args:
             action (Tensor): discrete actions with shape ``[batch_size]``.
-            q_values (Tensor): Q values with shape ``[batch_size, replicas, num_actions]``.
+            q_values (Tensor): Q values with shape
+                ``[batch_size, replicas, num_actions, reward_dim]``, where
+                ``reward_dim`` is optional for multi-dim reward.
         Returns:
-            Tensor: selected Q values with shape ``[batch_size, replicas]``.
+            Tensor: selected Q values with shape
+                ``[batch_size, replicas, reward_dim]``.
         """
-        # action shape: [batch_size] -> [batch_size, n, 1]
-        action = action.view(q_values.shape[0], 1, -1).expand(
-            -1, q_values.shape[1], -1).long()
-        return q_values.gather(-1, action).squeeze(-1)
+        ones = [1] * len(self._reward_spec.shape)
+        # [batch_size] -> [batch_size, 1, 1, ...]
+        action = action.view(q_values.shape[0], 1, 1, *ones)
+        # [batch_size, 1, 1, ...] -> [batch_size, n, 1, reward_dim]
+        action = action.expand(-1, q_values.shape[1], -1,
+                               *self._reward_spec.shape).long()
+        return q_values.gather(2, action).squeeze(2)
 
     def _critic_train_step(self, inputs: TimeStep, state: SacCriticState,
                            rollout_info: SacInfo, action, action_distribution):
         critics, critics_state = self._compute_critics(
-            self._critic_networks, inputs.observation, rollout_info.action,
-            state.critics)
+            self._critic_networks,
+            inputs.observation,
+            rollout_info.action,
+            state.critics,
+            replica_min=False,
+            apply_reward_weights=False)
 
         target_critics, target_critics_state = self._compute_critics(
-            self._target_critic_networks, inputs.observation, action,
-            state.target_critics)
-
-        if self.has_multidim_reward():
-            sign = self.reward_weights.sign()
-            target_critics = (target_critics * sign).min(dim=1)[0] * sign
-        else:
-            target_critics = target_critics.min(dim=1)[0]
+            self._target_critic_networks,
+            inputs.observation,
+            action,
+            state.target_critics,
+            apply_reward_weights=False)
 
         if self._act_type == ActionType.Discrete:
             critics = self._select_q_value(rollout_info.action, critics)
-            target_critics = self._select_q_value(
-                action, target_critics.unsqueeze(dim=1))
+            # [B, num_actions] -> [B, num_actions, reward_dim]
+            probs = common.expand_dims_as(action_distribution.probs,
+                                          target_critics)
+            # [B, reward_dim]
+            target_critics = torch.sum(probs * target_critics, dim=1)
```

Comment:
Changed from sampling estimate to weighted average. Now the target critics should be more accurate.